<div class="alert alert-success">
<h2>Комментарий ревьюера V.2</h2>

👍 Спасибо, что поправил и прокомментировал, удачи в следуюищх спринтах 
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера</h2>
    
Привет! Меня зовут Алексей Тюрин, я буду делать ревью твоей работы. Давай будем общаться на «ты».

У тебя хорошая, внимательная работа 👍. В целом всё круто, ничего исправлять не надо, я вернул работу, чтобы ты мог посмотреть комментарии и задать вопросы, если есть. Можно попробовать чуть хитрее заполнить прпоуски и определить "рисковость" каждой группы клиентов.


Зелёным цветом отмечены комментарии, формально не требующие правки, но правки приветствуются. Жёлтым цветом выделено то, что в следующий раз можно сделать по-другому или можно поправить прямо сейчас.

Давай работать над проектом в диалоге: если ты **что-то меняешь** в проекте по моим рекомендациям — **пиши об этом**. Выбери для своих комментариев какой-то заметный цвет, так мне будет легче отследить изменения. Пожалуйста, **не перемещай, не изменяй и не удаляй мои комментарии**. Всё это поможет выполнить повторную проверку твоего проекта оперативнее». Если оставляешь комментарии на какой-то итерации ревью, пиши что-то вроде `комментарий студента V.номер итерации`
</div>

Цвет можно сделать, например, так

<div class = "alert alert-info" >
<h2>Комментарий студента</h2>
Ревьюер душнила :)
</div>

Или так:
    
<div class="alert alert-info" style="background:pink;color:#A91E69">
<h2>Комментарий студента </h2>
    ...
</div>

<div class = "alert alert-info" >
<h2>Комментарий студента 1</h2>
Привет, Алексей!Меня зовут Сергей Холоденин, для меня это все в новинку и безумно интересно. Спасибо за твои комментарии, мне приятно, что ты оценил мою работу так высоко! Я доработал коды по твоим комментариям, буду рад обратной связи!
</div>

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## Шаг 1. Откройте файл с данными и изучите общую информацию

In [1]:
import pandas as pd
df = pd.read_csv('/datasets/data.csv')
df.head(15)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   

     total_income                            purpose  
0   253875.639453                      покупка жилья  
1   112080.014102            приобретение автомобиля  
2   145885.952297                      покупка жилья  
3   267628.550329         дополнительное образование  
4   158616.077870                    сыграть свадьбу  
5   255763.565419                      покупка жилья  
6   240525.971920                  операции с жильем  
7   135823.934197                        образование  
8    95856.832424              на проведение свадьбы  
9   144425.938277            покупка жилья для семьи  
10  113943.491460               покупка недвижимости  
11   77069.234271  покупка коммерческой недвижимости  
12            NaN                    сыграть свадьбу  
13  130458.228857            приобретение автомобиля  
14  165127.911772         покупка жилой недвижимости

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


**Вывод**

Из данных выше видно:
1. В столбцах days_employed и total_income очень много цифр после запятой, которые мешают анализу, следует заменить тип данных на целочисленный.
2. В столбце education используется различный регистр, следует привести все символы к строчным.
3. В столбцах days_employed и total_income есть пропущенные значения, следует проверить всю таблицу и учесть при анализе.
4. В столбце purpose похожие цели обозначены по разному, требуется категоризация, чтобы было возможно сгруппировать.
5. В столбце days_employed (трудовой стаж в днях) отрицательные значения, что является ошибкой, также есть сверхвысокие цифры, которые не отвечают реальности (возрасту заемщика).



<div class="alert alert-success">
<h2>Комментарий ревьюера</h2>

Хороший первый взгляд 👍
</div>

<div class = "alert alert-info" >
<h2>Комментарий студента 1</h2>
Спасибо!
</div>

## Шаг 2. Предобработка данных

### Обработка пропусков

In [3]:
df.isna().sum()


children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [4]:
df.groupby('income_type').agg({'days_employed': ['count', 'min', 'max','median', 'mean'], 'total_income': ['min','max','median','mean'], 'debt': ['sum']})

days_employed                                               \
                        count            min            max         median   
income_type                                                                  
безработный                 2  337524.466835  395302.838654  366413.652744   
в декрете                   1   -3296.759962   -3296.759962   -3296.759962   
госслужащий              1312  -15193.032201     -39.954170   -2689.368353   
компаньон                4577  -17615.563266     -30.195337   -1547.382223   
пенсионер                3443  328728.720605  401755.400475  365213.306266   
предприниматель             1    -520.848083    -520.848083    -520.848083   
сотрудник               10014  -18388.949901     -24.141633   -1574.202821   
студент                     1    -578.751554    -578.751554    -578.751554   

                                 total_income                               \
                          mean            min           max         median   
income_type                                                                  
безработный      366413.652744   59956.991984  2.027225e+05  131339.751676   
в декрете         -3296.759962   53829.130729  5.382913e+04   53829.130729   
госслужащий       -3399.896902   29200.077193  9.104515e+05  150447.935283   
компаньон         -2111.524398   28702.812889  2.265604e+06  172357.950966   
пенсионер        365003.491245   20667.263793  7.351033e+05  118514.486412   
предприниматель    -520.848083  499163.144947  4.991631e+05  499163.144947   
сотрудник         -2326.499216   21367.648356  1.726276e+06  142594.396847   
студент            -578.751554   98201.625314  9.820163e+04   98201.625314   

                                debt  
                          mean   sum  
income_type                           
безработный      131339.751676     1  
в декрете         53829.130729     1  
госслужащий      170898.309923    86  
компаньон        202417.461462   376  
пенсионер        137127.465690   216  
предприниматель  499163.144947     0  
сотрудник        161380.260488  1061  
студент           98201.625314     0

Пропуски есть в столбцах days_employed и total_income. Нужно их обработать. Отрицательные значения стажа - скорее всего ошибка при заполнении, хотя из сводной видно, что положительный стаж выше реально возможного и выставлен у категорий пенсионеры и безработные. Предположительно при заполнении вместо дней в данные были внесены часы стажа. Нужно откорректировать это, чтобы медианное занчение вычислялось корректно.

In [5]:
def days_employed(row):
    income_type=row['income_type']
    days_employed=row['days_employed']
    if income_type =='безработный':
        return days_employed/24
    if income_type =='пенсионер':
        return days_employed/24
    return days_employed
df['days_employed']=df.apply(days_employed,axis=1)
df.head(15)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0   14177.753002         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   

     total_income                            purpose  
0   253875.639453                      покупка жилья  
1   112080.014102            приобретение автомобиля  
2   145885.952297                      покупка жилья  
3   267628.550329         дополнительное образование  
4   158616.077870                    сыграть свадьбу  
5   255763.565419                      покупка жилья  
6   240525.971920                  операции с жильем  
7   135823.934197                        образование  
8    95856.832424              на проведение свадьбы  
9   144425.938277            покупка жилья для семьи  
10  113943.491460               покупка недвижимости  
11   77069.234271  покупка коммерческой недвижимости  
12            NaN                    сыграть свадьбу  
13  130458.228857            приобретение автомобиля  
14  165127.911772         покупка жилой недвижимости

<div class="alert alert-success">
<h2>Комментарий ревьюера</h2>

👍 Можно было в принципе без функции просто срез заполнить: `df.loc[(df['income_type'] == 'безработный') | (df['income_type'] == 'пенсионер')] = df[(df['income_type'] == 'безработный') | (df['income_type'] == 'пенсионер')] / 24 `
</div>

<div class = "alert alert-info" >
<h2>Комментарий студента 1</h2>
Здесь не стану менять код, возьму на заметку, мне интересно потренироваться по написанию функций :)
</div>

Теперь можно привести все по модулю в столбце days_employed и вычислить медиану, т.к. она более точно отражает средний стаж. После чего пропуски заменяю медианой в этом столбце.

Теперь можно по медианному значению отфильтровать таблицу и посмотреть информацию по пропускам. Изначально пропусков было 2174, после обработки информации стажа по пенсионерам осталось 1761 пропуск. Пропускав столбцах days_employed и total_income полностью соответсвуют. Из чего можно предположить, что по данным клиентам не было предоставлено сведений о работе, поэтому не известен стаж и не известно суммы дохода.

Теперь по медианному значению можно посмотреть, совпадают ли пропуски с со стотбцом total_income, т.к. количество значений совпадает. Из данных видно, что теория имеет место быть.

<div class="alert alert-success">
<h2>Комментарий ревьюера</h2>

Так как стаж вероятнее всего зависит от возраста, то можно разбить на группы по возрасту и заменить на медиану уже по группам, вероятность ошибочных значений в таком случае была бы меньше. В этом помогут методы `groupby` + `transform` + `fillna`. Ещё тут есть сьолкт с нулевым возрастом, их следует либо заполнить либо удалить, перед тем как проверять на ошибки
</div>

<div class = "alert alert-info" >
<h2>Комментарий студента 1</h2>
Последовал твоему совету и переработал данный блок анализа согласно твоим рекомендациям:
</div>

<div class="alert alert-success">
<h2>Комментарий ревьюера V.2</h2>

👍
</div>

Сперва проверим какие возраста есть в выборке:

In [6]:
df.groupby('dob_years')['dob_years'].count()

dob_years
0     101
19     14
20     51
21    111
22    183
23    254
24    264
25    357
26    408
27    493
28    503
29    545
30    540
31    560
32    510
33    581
34    603
35    617
36    555
37    537
38    598
39    573
40    609
41    607
42    597
43    513
44    547
45    497
46    475
47    480
48    538
49    508
50    514
51    448
52    484
53    459
54    479
55    443
56    487
57    460
58    461
59    444
60    377
61    355
62    352
63    269
64    265
65    194
66    183
67    167
68     99
69     85
70     65
71     58
72     33
73      8
74      6
75      1
Name: dob_years, dtype: int64

Значения 0 в возрасте - явная ошибка, заменяю на медианное значение данные пропуска:

In [7]:
dob_years_median = df['dob_years'].median()
df['dob_years']=df['dob_years'].replace(0, dob_years_median)
df.groupby('dob_years')['dob_years'].count()

dob_years
19     14
20     51
21    111
22    183
23    254
24    264
25    357
26    408
27    493
28    503
29    545
30    540
31    560
32    510
33    581
34    603
35    617
36    555
37    537
38    598
39    573
40    609
41    607
42    698
43    513
44    547
45    497
46    475
47    480
48    538
49    508
50    514
51    448
52    484
53    459
54    479
55    443
56    487
57    460
58    461
59    444
60    377
61    355
62    352
63    269
64    265
65    194
66    183
67    167
68     99
69     85
70     65
71     58
72     33
73      8
74      6
75      1
Name: dob_years, dtype: int64

Для того, чтобы корректней вставить медианное значение, разбиваю на возрастные группы с шагом 10 лет до 65, далее пенсионеры.

In [8]:
def years_group(years):
    if years <20:
        return 'до 20'
    if years <30:
        return '20-30'
    if years <40:
        return '30-40'
    if years <50:
        return '40-50'
    if years <60:
        return '50-60'
    if years <65:
        return '60-65'
    return 'старше 65'  
df['years_group']=df['dob_years'].apply(years_group)
df.head()

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0   14177.753002         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose years_group  
0               покупка жилья       40-50  
1     приобретение автомобиля       30-40  
2               покупка жилья       30-40  
3  дополнительное образование       30-40  
4             сыграть свадьбу       50-60

По каждой из групп считаю медианное значение и заменяю им пропуска в столбце days_employed. Предварительно отрицательные значения привожу по модулю.

In [9]:
df['days_employed']=df['days_employed'].abs()
df['median_employed']=df.groupby('years_group')['days_employed'].transform('median')
df['days_employed']=df['days_employed'].fillna(df['median_employed'])
df.head(15)


children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42               высшее             0   
1          1    4024.803754         36              среднее             1   
2          0    5623.422610         33              Среднее             1   
3          3    4124.747207         32              среднее             1   
4          0   14177.753002         53              среднее             1   
5          0     926.185831         27               высшее             0   
6          0    2879.202052         43               высшее             0   
7          0     152.779569         50              СРЕДНЕЕ             1   
8          2    6929.865299         35               ВЫСШЕЕ             0   
9          0    2188.756445         41              среднее             1   
10         2    4171.483647         36               высшее             0   
11         0     792.701887         40              среднее             1   
12         0   15020.871042         65              среднее             1   
13         0    1846.641941         54  неоконченное высшее             2   
14         0    1844.956182         56               высшее             0   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   

     total_income                            purpose years_group  \
0   253875.639453                      покупка жилья       40-50   
1   112080.014102            приобретение автомобиля       30-40   
2   145885.952297                      покупка жилья       30-40   
3   267628.550329         дополнительное образование       30-40   
4   158616.077870                    сыграть свадьбу       50-60   
5   255763.565419                      покупка жилья       20-30   
6   240525.971920                  операции с жильем       40-50   
7   135823.934197                        образование       50-60   
8    95856.832424              на проведение свадьбы       30-40   
9   144425.938277            покупка жилья для семьи       40-50   
10  113943.491460               покупка недвижимости       30-40   
11   77069.234271  покупка коммерческой недвижимости       40-50   
12            NaN                    сыграть свадьбу   старше 65   
13  130458.228857            приобретение автомобиля       50-60   
14  165127.911772         покупка жилой недвижимости       50-60   

    median_employed  
0       2108.529344  
1       1601.784231  
2       1601.784231  
3       1601.784231  
4       4796.767897  
5       1005.629955  
6       2108.529344  
7       4796.767897  
8       1601.784231  
9       2108.529344  
10      1601.784231  
11      2108.529344  
12     15020.871042  
13      4796.767897  
14      4796.767897

Теперь можно по медианному значению отфильтровать таблицу и посмотреть информацию по пропускам.  Пропуска в столбцах days_employed и total_income соответсвуют (2174 значений). Из чего можно предположить, что по данным клиентам не было предоставлено сведений о работе, поэтому не известен стаж и не известно суммы дохода.

In [10]:
median_table=df[df['days_employed']== df['median_employed']]
median_table.groupby('income_type').agg({'days_employed': ['count', 'min', 'max','median', 'mean'], 'total_income': ['min','max','median','mean'], 'debt': ['sum']})

days_employed                                          \
                        count          min           max       median   
income_type                                                             
госслужащий               147  1005.629955  14649.209528  2108.529344   
компаньон                 509   724.492610  15020.871042  2108.529344   
пенсионер                 414  1005.629955  15020.871042  4796.767897   
предприниматель             1  4796.767897   4796.767897  4796.767897   
сотрудник                1108   724.492610  15020.871042  2108.529344   

                               total_income                                \
                        mean            min            max         median   
income_type                                                                 
госслужащий      2905.509802            NaN            NaN            NaN   
компаньон        2613.270462   59124.071006   59124.071006   59124.071006   
пенсионер        9453.721086  175186.979393  175186.979393  175186.979393   
предприниматель  4796.767897            NaN            NaN            NaN   
сотрудник        2684.424005   66592.548107   99890.305894   93403.748031   

                               debt  
                          mean  sum  
income_type                          
госслужащий                NaN    7  
компаньон         59124.071006   30  
пенсионер        175186.979393   35  
предприниматель            NaN    0  
сотрудник         86628.867344   98

Теперь по медианному значению можно посмотреть, совпадают ли пропуски с со стотбцом total_income, т.к. количество значений совпадает. Из данных видно, что теория имеет место быть.

In [11]:
median_table.isna().sum()

children               0
days_employed          0
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
years_group            0
median_employed        0
dtype: int64

In [12]:
median_table.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2179 entries, 12 to 21510
Data columns (total 14 columns):
children            2179 non-null int64
days_employed       2179 non-null float64
dob_years           2179 non-null int64
education           2179 non-null object
education_id        2179 non-null int64
family_status       2179 non-null object
family_status_id    2179 non-null int64
gender              2179 non-null object
income_type         2179 non-null object
debt                2179 non-null int64
total_income        5 non-null float64
purpose             2179 non-null object
years_group         2179 non-null object
median_employed     2179 non-null float64
dtypes: float64(3), int64(5), object(6)
memory usage: 335.4+ KB


Нужно проверить на корректность остальных строк стажа. Ошибка заполнения может влиять на результат анализа. Для этого возраст за вычетом 18 лет (совершеннолетие и начало трудовой деятельности) я умножаю на 365 дней, после чего вычитаю количество дней стажа. При отрицательном занчении в столбце days_employed_error отражается 1 (т.е. это ошибочные данные).

In [13]:
def days_employed_error(row):
    dob_years=row['dob_years']
    days_employed=row['days_employed']
    if (dob_years-18)*365-days_employed <0:
        return 1
    return 0
df['days_employed_error']=df.apply(days_employed_error, axis=1)
df.head(15)

children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42               высшее             0   
1          1    4024.803754         36              среднее             1   
2          0    5623.422610         33              Среднее             1   
3          3    4124.747207         32              среднее             1   
4          0   14177.753002         53              среднее             1   
5          0     926.185831         27               высшее             0   
6          0    2879.202052         43               высшее             0   
7          0     152.779569         50              СРЕДНЕЕ             1   
8          2    6929.865299         35               ВЫСШЕЕ             0   
9          0    2188.756445         41              среднее             1   
10         2    4171.483647         36               высшее             0   
11         0     792.701887         40              среднее             1   
12         0   15020.871042         65              среднее             1   
13         0    1846.641941         54  неоконченное высшее             2   
14         0    1844.956182         56               высшее             0   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   

     total_income                            purpose years_group  \
0   253875.639453                      покупка жилья       40-50   
1   112080.014102            приобретение автомобиля       30-40   
2   145885.952297                      покупка жилья       30-40   
3   267628.550329         дополнительное образование       30-40   
4   158616.077870                    сыграть свадьбу       50-60   
5   255763.565419                      покупка жилья       20-30   
6   240525.971920                  операции с жильем       40-50   
7   135823.934197                        образование       50-60   
8    95856.832424              на проведение свадьбы       30-40   
9   144425.938277            покупка жилья для семьи       40-50   
10  113943.491460               покупка недвижимости       30-40   
11   77069.234271  покупка коммерческой недвижимости       40-50   
12            NaN                    сыграть свадьбу   старше 65   
13  130458.228857            приобретение автомобиля       50-60   
14  165127.911772         покупка жилой недвижимости       50-60   

    median_employed  days_employed_error  
0       2108.529344                    0  
1       1601.784231                    0  
2       1601.784231                    1  
3       1601.784231                    0  
4       4796.767897                    1  
5       1005.629955                    0  
6       2108.529344                    0  
7       4796.767897                    0  
8       1601.784231                    1  
9       2108.529344                    0  
10      1601.784231                    0  
11      2108.529344                    0  
12     15020.871042                    0  
13     

Посчитаем какова доля строк с ошибочным стажем в days_employed:

In [14]:
df['days_employed_error'].sum()/df['days_employed_error'].count()

0.09955865272938444

10% данных по стажу некорректно заполненj, что частично влияет на общий анализ, но не критично.

Теперь заменим пропуски в столбце total_income. Сперва произведем группировку income_type по уровню дохода для того, чтобы медианные значения были более корректными. В пропущенных значениях income_type ставлю низкий уровень дохода.

In [15]:
df['income_type']=pd.qcut(df['total_income'],
                          q=[0, 0.2, .4, .6, .8, 1],
                          labels=['низкий','ниже среднего','средний','выше среднего','высокий',])
df['income_type']=df['income_type'].fillna('низкий')

По аналогии с days_employed провожу замену пропусков медианным значением в зависимости от уровня доходов.

In [16]:
df['median_income']=df.groupby('income_type')['total_income'].transform('median')
df['total_income']=df['total_income'].fillna(df['median_income'])
df.head(15)

children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42               высшее             0   
1          1    4024.803754         36              среднее             1   
2          0    5623.422610         33              Среднее             1   
3          3    4124.747207         32              среднее             1   
4          0   14177.753002         53              среднее             1   
5          0     926.185831         27               высшее             0   
6          0    2879.202052         43               высшее             0   
7          0     152.779569         50              СРЕДНЕЕ             1   
8          2    6929.865299         35               ВЫСШЕЕ             0   
9          0    2188.756445         41              среднее             1   
10         2    4171.483647         36               высшее             0   
11         0     792.701887         40              среднее             1   
12         0   15020.871042         65              среднее             1   
13         0    1846.641941         54  неоконченное высшее             2   
14         0    1844.956182         56               высшее             0   

       family_status  family_status_id gender    income_type  debt  \
0    женат / замужем                 0      F        высокий     0   
1    женат / замужем                 0      F  ниже среднего     0   
2    женат / замужем                 0      M        средний     0   
3    женат / замужем                 0      M        высокий     0   
4   гражданский брак                 1      F        средний     0   
5   гражданский брак                 1      M        высокий     0   
6    женат / замужем                 0      F        высокий     0   
7    женат / замужем                 0      M        средний     0   
8   гражданский брак                 1      F  ниже среднего     0   
9    женат / замужем                 0      M        средний     0   
10   женат / замужем                 0      M  ниже среднего     0   
11   женат / замужем                 0      F         низкий     0   
12  гражданский брак                 1      M         низкий     0   
13   женат / замужем                 0      F        средний     0   
14  гражданский брак                 1      F  выше среднего     1   

     total_income                            purpose years_group  \
0   253875.639453                      покупка жилья       40-50   
1   112080.014102            приобретение автомобиля       30-40   
2   145885.952297                      покупка жилья       30-40   
3   267628.550329         дополнительное образование       30-40   
4   158616.077870                    сыграть свадьбу       50-60   
5   255763.565419                      покупка жилья       20-30   
6   240525.971920                  операции с жильем       40-50   
7   135823.934197                        образование       50-60   
8    95856.832424              на проведение свадьбы       30-40   
9   144425.938277            покупка жилья для семьи       40-50   
10  113943.491460               покупка недвижимости       30-40   
11   77069.234271  покупка коммерческой недвижимости       40-50   
12   76095.095134                    сыграть свадьбу   старше 65   
13  130458.228857            приобретение автомобиля       50-60   
14  165127.911772         покупка жилой недвижимости       50-60   

    median_employed  days_employed_error  median_income  
0       2108.529344                    0  278423.213130  
1       1601.784231                    0  111206.959153  
2       1601.784231                    1  145020.801280  
3       1601.784231                    0  278423.213130  
4       4796.767897                    1  145020.801280  
5       1005.629955                    0  278423.213130  
6       2108.529344                    0  278423.213130  
7       4796.767897                    0  145020.801280  
8       1601.784231                    

<div class="alert alert-success">
<h2>Комментарий ревьюера</h2>

Здесь аналогично, это неплохая замена, но она сместит средний доход в каждой группе, т.к. медианный может быть выше чем доход в группе и создаст большой пик в медиане, что повлияет на корреляцию и дисперсию. Опять же можно попробовать заполнить прпоуски по группам, в зависимости от `income_type`
</div>

<div class = "alert alert-info" >
<h2>Комментарий студента 1</h2>
Произвел корректировку заполнения пропусков в доходе аналогично стажу.
</div>

<div class="alert alert-success">
<h2>Комментарий ревьюера V.2</h2>

👍 Хороший способ, но `income_type` уже был :) Это так называлась профессия
</div>

In [17]:
df.groupby('income_type').agg({'days_employed': ['count', 'min', 'max','median', 'mean'], 'total_income': ['min','max','median','mean'], 'debt': ['sum']})

days_employed                                        \
                      count        min           max       median   
income_type                                                         
низкий                 6045  34.701045  16738.158823  2108.529344   
ниже среднего          3870  24.240695  17615.563266  2141.513698   
средний                3870  33.520665  16733.936484  2017.665522   
выше среднего          3870  24.141633  18388.949901  2026.045016   
высокий                3870  30.195337  16736.436110  2113.007971   

                             total_income                               \
                      mean            min           max         median   
income_type                                                              
низкий         5394.733645   20667.263793  9.483683e+04   76095.095134   
ниже среднего  4797.805321   94841.530758  1.282797e+05  111206.959153   
средний        4332.678787  128288.439652  1.646609e+05  145020.801280   
выше среднего  4117.055372  164664.737222  2.231076e+05  189092.626371   
высокий        3765.222165  223175.713391  2.265604e+06  278423.213130   

                             debt  
                        mean  sum  
income_type                        
низкий          74186.832426  479  
ниже среднего  111266.819775  324  
средний        145784.359202  335  
выше среднего  190268.497880  329  
высокий        316701.073748  274

In [18]:
df.isna().sum()

children               0
days_employed          0
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
years_group            0
median_employed        0
days_employed_error    0
median_income          0
dtype: int64

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 16 columns):
children               21525 non-null int64
days_employed          21525 non-null float64
dob_years              21525 non-null int64
education              21525 non-null object
education_id           21525 non-null int64
family_status          21525 non-null object
family_status_id       21525 non-null int64
gender                 21525 non-null object
income_type            21525 non-null category
debt                   21525 non-null int64
total_income           21525 non-null float64
purpose                21525 non-null object
years_group            21525 non-null object
median_employed        21525 non-null float64
days_employed_error    21525 non-null int64
median_income          21525 non-null float64
dtypes: category(1), float64(4), int64(6), object(5)
memory usage: 2.5+ MB


**Вывод**

Проанализировав информацию по пропускам были выявлены столбцы days_employed и total_income с пропусками. Пропуск в них связан со сведениями о работе: стаж и доход, соответственно данные клиенты не предоставили информации по своему месту работы (стаж и доход). Для того, чтобы не терять данные по этим клиентам, пропуска заполнены медианным значением в каждом из столбцов.

### Замена типа данных

В столбцах days_employed и total_income очень много цифр после запятой, которые мешают анализу, следует заменить тип данных на целочисленный.

In [20]:
df['days_employed']=df['days_employed'].astype(int)
df['total_income']=df['total_income'].astype(int)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 16 columns):
children               21525 non-null int64
days_employed          21525 non-null int64
dob_years              21525 non-null int64
education              21525 non-null object
education_id           21525 non-null int64
family_status          21525 non-null object
family_status_id       21525 non-null int64
gender                 21525 non-null object
income_type            21525 non-null category
debt                   21525 non-null int64
total_income           21525 non-null int64
purpose                21525 non-null object
years_group            21525 non-null object
median_employed        21525 non-null float64
days_employed_error    21525 non-null int64
median_income          21525 non-null float64
dtypes: category(1), float64(2), int64(8), object(5)
memory usage: 2.5+ MB


In [21]:
df.head(15)

children  days_employed  dob_years            education  education_id  \
0          1           8437         42               высшее             0   
1          1           4024         36              среднее             1   
2          0           5623         33              Среднее             1   
3          3           4124         32              среднее             1   
4          0          14177         53              среднее             1   
5          0            926         27               высшее             0   
6          0           2879         43               высшее             0   
7          0            152         50              СРЕДНЕЕ             1   
8          2           6929         35               ВЫСШЕЕ             0   
9          0           2188         41              среднее             1   
10         2           4171         36               высшее             0   
11         0            792         40              среднее             1   
12         0          15020         65              среднее             1   
13         0           1846         54  неоконченное высшее             2   
14         0           1844         56               высшее             0   

       family_status  family_status_id gender    income_type  debt  \
0    женат / замужем                 0      F        высокий     0   
1    женат / замужем                 0      F  ниже среднего     0   
2    женат / замужем                 0      M        средний     0   
3    женат / замужем                 0      M        высокий     0   
4   гражданский брак                 1      F        средний     0   
5   гражданский брак                 1      M        высокий     0   
6    женат / замужем                 0      F        высокий     0   
7    женат / замужем                 0      M        средний     0   
8   гражданский брак                 1      F  ниже среднего     0   
9    женат / замужем                 0      M        средний     0   
10   женат / замужем                 0      M  ниже среднего     0   
11   женат / замужем                 0      F         низкий     0   
12  гражданский брак                 1      M         низкий     0   
13   женат / замужем                 0      F        средний     0   
14  гражданский брак                 1      F  выше среднего     1   

    total_income                            purpose years_group  \
0         253875                      покупка жилья       40-50   
1         112080            приобретение автомобиля       30-40   
2         145885                      покупка жилья       30-40   
3         267628         дополнительное образование       30-40   
4         158616                    сыграть свадьбу       50-60   
5         255763                      покупка жилья       20-30   
6         240525                  операции с жильем       40-50   
7         135823                        образование       50-60   
8          95856              на проведение свадьбы       30-40   
9         144425            покупка жилья для семьи       40-50   
10        113943               покупка недвижимости       30-40   
11         77069  покупка коммерческой недвижимости       40-50   
12         76095                    сыграть свадьбу   старше 65   
13        130458            приобретение автомобиля       50-60   
14        165127         покупка жилой недвижимости       50-60   

    median_employed  days_employed_error  median_income  
0       2108.529344                    0  278423.213130  
1       1601.784231                    0  111206.959153  
2       1601.784231                    1  145020.801280  
3       1601.784231                    0  278423.213130  
4       4796.767897                    1  145020.801280  
5       1005.629955                    0  278423.213130  
6       2108.529344                    0  278423.213130  
7       4796.767897                    0  145020.801280  
8       1601.784231                    1  111206.959153

**Вывод**

После преобразования таблица стала более доступна для восприятия и анализа.

### Обработка дубликатов

Проверим наличие дубликатов методом duplicated() и просуммируем sum().

In [22]:
df.duplicated().sum()

101

101 строка - дубликаты, уберем дубли методом drop_duplicates() и проверим их наличие после. Дубли могли возникнуть при повторной заявке клиента. 

In [23]:
df = df.drop_duplicates().reset_index(drop=True)
df.duplicated().sum()

0

**Вывод**

В таблице присутствовало небольшое количество дубликатов, убрав их можно продолжать исследование.

### Лемматизация

Перед тем, как заняться лемматизацией , следует привести все символы к строчным в столбце education. 

In [24]:
df['education']=df['education'].str.lower()
df.head(15)

children  days_employed  dob_years            education  education_id  \
0          1           8437         42               высшее             0   
1          1           4024         36              среднее             1   
2          0           5623         33              среднее             1   
3          3           4124         32              среднее             1   
4          0          14177         53              среднее             1   
5          0            926         27               высшее             0   
6          0           2879         43               высшее             0   
7          0            152         50              среднее             1   
8          2           6929         35               высшее             0   
9          0           2188         41              среднее             1   
10         2           4171         36               высшее             0   
11         0            792         40              среднее             1   
12         0          15020         65              среднее             1   
13         0           1846         54  неоконченное высшее             2   
14         0           1844         56               высшее             0   

       family_status  family_status_id gender    income_type  debt  \
0    женат / замужем                 0      F        высокий     0   
1    женат / замужем                 0      F  ниже среднего     0   
2    женат / замужем                 0      M        средний     0   
3    женат / замужем                 0      M        высокий     0   
4   гражданский брак                 1      F        средний     0   
5   гражданский брак                 1      M        высокий     0   
6    женат / замужем                 0      F        высокий     0   
7    женат / замужем                 0      M        средний     0   
8   гражданский брак                 1      F  ниже среднего     0   
9    женат / замужем                 0      M        средний     0   
10   женат / замужем                 0      M  ниже среднего     0   
11   женат / замужем                 0      F         низкий     0   
12  гражданский брак                 1      M         низкий     0   
13   женат / замужем                 0      F        средний     0   
14  гражданский брак                 1      F  выше среднего     1   

    total_income                            purpose years_group  \
0         253875                      покупка жилья       40-50   
1         112080            приобретение автомобиля       30-40   
2         145885                      покупка жилья       30-40   
3         267628         дополнительное образование       30-40   
4         158616                    сыграть свадьбу       50-60   
5         255763                      покупка жилья       20-30   
6         240525                  операции с жильем       40-50   
7         135823                        образование       50-60   
8          95856              на проведение свадьбы       30-40   
9         144425            покупка жилья для семьи       40-50   
10        113943               покупка недвижимости       30-40   
11         77069  покупка коммерческой недвижимости       40-50   
12         76095                    сыграть свадьбу   старше 65   
13        130458            приобретение автомобиля       50-60   
14        165127         покупка жилой недвижимости       50-60   

    median_employed  days_employed_error  median_income  
0       2108.529344                    0  278423.213130  
1       1601.784231                    0  111206.959153  
2       1601.784231                    1  145020.801280  
3       1601.784231                    0  278423.213130  
4       4796.767897                    1  145020.801280  
5       1005.629955                    0  278423.213130  
6       2108.529344                    0  278423.213130  
7       4796.767897                    0  145020.801280  
8       1601.784231                    1  111206.959153

Лемматизацию буду проводить по столбцу purpose, это позволит провести категоризацию по целям. Как видно, много повторяющихся целей, которые записаны в разной формулировке:

In [25]:
df.groupby('purpose')['purpose'].count()

purpose
автомобили                                477
автомобиль                                494
высшее образование                        451
дополнительное образование                460
жилье                                     643
заняться высшим образованием              495
заняться образованием                     407
на покупку автомобиля                     472
на покупку подержанного автомобиля        478
на покупку своего автомобиля              505
на проведение свадьбы                     768
недвижимость                              630
образование                               447
операции с жильем                         652
операции с коммерческой недвижимостью     649
операции с недвижимостью                  675
операции со своей недвижимостью           626
покупка жилой недвижимости                604
покупка жилья                             646
покупка жилья для сдачи                   652
покупка жилья для семьи                   636
покупка коммерческой недви

Проведем лемматизацию и посчитаем количество повторов начальной формы слова, исходя из данных выделим категории.

<div class="alert alert-success">
<h2>Комментарий ревьюера</h2>

Код лучше логически разбивать на блоки через новую строку. Ставить пробелы вокруг `=`, а импорты иметь в самом верху тетрадки. Вообще если писать в тетрадке в какой-нибудь среде разработки (VS Code например), то если наставить расширений, то оно само будет об этом говорить
    
    
    
Чтобы дубликаты не удалять, есть метод `unique()`. Ещё кажется что может быть важно сохранять контекст слова, т.е. чтобы понимать, когда жильё, а когда ремонт. Тут не очень много целей, но на большом списке это было бы полезно. Т.е. можно для каждой цели из `df['purpose'].unique()` сначала её лемматизировать, потом убрать предлоги и `\n`, а потом соединить обратно в строку через `' '.join(lemma)` тогда бы итоговый список сохранил контекст
</div>

<div class = "alert alert-info" >
<h2>Комментарий студента 1</h2>
Принял в работу, с лемматизацией попробую поиграться отдельно, здесь менять не стал. Спасибо за наставление.
</div>

In [26]:

queries_t=df['purpose']
queries_t=queries_t.drop_duplicates().reset_index(drop= True)
queries=queries_t.tolist()
from pymystem3 import Mystem
m = Mystem()
words=[' ']
for word in queries:
    lemmas=m.lemmatize(word)
    
    words += lemmas
from collections import Counter
Counter(words)

   




Counter({' ': 60,
         'покупка': 10,
         'жилье': 7,
         '\n': 38,
         'приобретение': 1,
         'автомобиль': 9,
         'дополнительный': 2,
         'образование': 9,
         'сыграть': 1,
         'свадьба': 3,
         'операция': 4,
         'с': 5,
         'на': 4,
         'проведение': 1,
         'для': 2,
         'семья': 1,
         'недвижимость': 10,
         'коммерческий': 2,
         'жилой': 2,
         'строительство': 3,
         'собственный': 1,
         'подержать': 1,
         'свой': 4,
         'со': 1,
         'заниматься': 2,
         'сделка': 2,
         'подержанный': 1,
         'получение': 3,
         'высокий': 3,
         'профильный': 1,
         'сдача': 1,
         'ремонт': 1})

**Вывод**

Основные цели выглядят так:'жилье': 7,'автомобиль': 9,'образование': 9,'свадьба': 3,'недвижимость': 10, 'жилой': 2,'строительство': 3,'ремонт': 1.
При этом жилье, недвижимость, жилой, строительство, ремонт можно объединить в одну категорию: недвижимость.
Итоговые категории: 
-недвижимость
-автомобиль
-образование
-свадьба

### Категоризация данных

Теперь нужно в столбце purpose заменить цели на корректные. Для этого нужно написать функцию.

In [27]:
from nltk.stem import SnowballStemmer 
russian_stemmer = SnowballStemmer('russian')
russian_stemmer.stem('автомобиль')

            
         

'автомобил'

In [28]:
russian_stemmer.stem('образование')

'образован'

In [29]:
russian_stemmer.stem('свадьба')

'свадьб'

In [30]:
def category_purpose(purpose):    
            if purpose.find('автомобил')>0:
                return 'автомобиль'
            if purpose.find('образован')>0:
                return 'образование'
            if purpose.find('свадьб')>0:
                return 'свадьба'
            return 'недвижимость'
df['category_purpose']=df['purpose'].apply(category_purpose)
df.head(15)

children  days_employed  dob_years            education  education_id  \
0          1           8437         42               высшее             0   
1          1           4024         36              среднее             1   
2          0           5623         33              среднее             1   
3          3           4124         32              среднее             1   
4          0          14177         53              среднее             1   
5          0            926         27               высшее             0   
6          0           2879         43               высшее             0   
7          0            152         50              среднее             1   
8          2           6929         35               высшее             0   
9          0           2188         41              среднее             1   
10         2           4171         36               высшее             0   
11         0            792         40              среднее             1   
12         0          15020         65              среднее             1   
13         0           1846         54  неоконченное высшее             2   
14         0           1844         56               высшее             0   

       family_status  family_status_id gender    income_type  debt  \
0    женат / замужем                 0      F        высокий     0   
1    женат / замужем                 0      F  ниже среднего     0   
2    женат / замужем                 0      M        средний     0   
3    женат / замужем                 0      M        высокий     0   
4   гражданский брак                 1      F        средний     0   
5   гражданский брак                 1      M        высокий     0   
6    женат / замужем                 0      F        высокий     0   
7    женат / замужем                 0      M        средний     0   
8   гражданский брак                 1      F  ниже среднего     0   
9    женат / замужем                 0      M        средний     0   
10   женат / замужем                 0      M  ниже среднего     0   
11   женат / замужем                 0      F         низкий     0   
12  гражданский брак                 1      M         низкий     0   
13   женат / замужем                 0      F        средний     0   
14  гражданский брак                 1      F  выше среднего     1   

    total_income                            purpose years_group  \
0         253875                      покупка жилья       40-50   
1         112080            приобретение автомобиля       30-40   
2         145885                      покупка жилья       30-40   
3         267628         дополнительное образование       30-40   
4         158616                    сыграть свадьбу       50-60   
5         255763                      покупка жилья       20-30   
6         240525                  операции с жильем       40-50   
7         135823                        образование       50-60   
8          95856              на проведение свадьбы       30-40   
9         144425            покупка жилья для семьи       40-50   
10        113943               покупка недвижимости       30-40   
11         77069  покупка коммерческой недвижимости       40-50   
12         76095                    сыграть свадьбу   старше 65   
13        130458            приобретение автомобиля       50-60   
14        165127         покупка жилой недвижимости       50-60   

    median_employed  days_employed_error  median_income category_purpose  
0       2108.529344                    0  278423.213130     недвижимость  
1       1601.784231                    0  111206.959153       автомобиль  
2       1601.784231                    1  145020.801280     недвижимость  
3       1601.784231                    0  278423.213130      образование  
4       4796.767897                    1  145020.801280          свадьба  
5       1005.629955                    0  278423.213130     недвижимость  
6       2108.529344                    0  278423.213

<div class="alert alert-success">
<h2>Комментарий ревьюера</h2>

👍
</div>

In [31]:
category_purpose_group=df.groupby('category_purpose').agg({ 'debt': ['sum','count','mean']})
category_purpose_group.columns=['debt','total','%']
debt_total = category_purpose_group['debt'].sum() / category_purpose_group['total'].sum()
def risk_group(percent):
    if percent >= debt_total:
        return 'yes'
    return 'no'
category_purpose_group['risk_group'] = category_purpose_group['%'].apply(risk_group)
category_purpose_group


debt  total         % risk_group
category_purpose                                  
автомобиль         317   3331  0.095167        yes
недвижимость       964  12998  0.074165         no
образование        338   3559  0.094970        yes
свадьба            122   1536  0.079427         no

Категории целей успешно созданы сгруппированы, также вычислены группы риска. теперь нужно сгруппировать для каждого анализа свой список и категоризировать.

In [32]:
df.groupby('children').agg({ 'debt' : ['sum' , 'count']})

debt       
           sum  count
children             
-1           1     47
 0        1063  14066
 1         444   4805
 2         194   2050
 3          27    330
 4           4     41
 5           0      9
 20          8     76

-1 и 20 детей - это явные ошибки ввода, приведем таблицу к -1=0, а 20=2, т.к. это скорее всего опечатки.

In [33]:
def children_change(child):
    if child == 20:
        return 2
    if child == -1:
        return 0
    return child

df['children']=df['children'].apply(children_change)
children_group=df.groupby('children').agg({ 'debt' : ['sum' , 'count','mean']})
children_group.columns=['debt','total','%']
children_group['risk_group'] = children_group['%'].apply(risk_group)
children_group

debt  total         % risk_group
children                                  
0         1064  14113  0.075391         no
1          444   4805  0.092404        yes
2          202   2126  0.095014        yes
3           27    330  0.081818        yes
4            4     41  0.097561        yes
5            0      9  0.000000         no

<div class="alert alert-success">
<h2>Комментарий ревьюера</h2>

В таких случаях ещё можно пользоваться функцией `replace`
</div>

<div class = "alert alert-info" >
<h2>Комментарий студента 1</h2>
Да, я просто решил потренироваться в написании функций. спасибо за совет!
</div>

Теперь группировка по количеству детей выглядит корректно. Теперь сделаем группировку по семейному положению.

In [34]:
family_status_group=df.groupby('family_status').agg({ 'debt' : ['sum' , 'count','mean']})
family_status_group.columns=['debt','total','%']
family_status_group['risk_group'] = family_status_group['%'].apply(risk_group)
family_status_group

debt  total         % risk_group
family_status                                          
Не женат / не замужем   274   2809  0.097544        yes
в разводе                85   1194  0.071189         no
вдовец / вдова           63    957  0.065831         no
гражданский брак        388   4154  0.093404        yes
женат / замужем         931  12310  0.075630         no

Теперь произведем группировку по уровню дохода.

In [35]:
income_type_group=df.groupby('income_type').agg({'debt' : ['sum' , 'count','mean']})
income_type_group.columns=['debt','total','%']
income_type_group['risk_group'] = income_type_group['%'].apply(risk_group)
income_type_group

debt  total         % risk_group
income_type                                    
низкий          479   5944  0.080585         no
ниже среднего   324   3870  0.083721        yes
средний         335   3870  0.086563        yes
выше среднего   329   3870  0.085013        yes
высокий         274   3870  0.070801         no

<div class="alert alert-success">
<h2>Комментарий ревьюера</h2>

👍 Классный способ :)
</div>

<div class = "alert alert-info" >
<h2>Комментарий студента 1</h2>
Это да, я его перенес в раздел с пропусками, чтобы по уровню дохода найти медианные значения и заменить пропуска=)
</div>

**Вывод**

Все данные сгруппированы исходя из тех вопросов, на которые требуется дать ответ. Таблицы сгруппированы по принципу суммирования и подсчета количества наблюдений в тех категориях, которые требуются для ответа на вопрос. Исходя из корреляции при выделении % я сделаю вывод, есть ли связь. Категоризация выполнена.

## Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [36]:
children_group =children_group.sort_values(by=['total'], ascending=False)
children_group

debt  total         % risk_group
children                                  
0         1064  14113  0.075391         no
1          444   4805  0.092404        yes
2          202   2126  0.095014        yes
3           27    330  0.081818        yes
4            4     41  0.097561        yes
5            0      9  0.000000         no

**Вывод**

Исходя из цифр выше видно, что при отсутствии детей процент не возврата ниже (7,5%), при наличии детей процент возрастает 9,2-9,7%, 9 наблюдений с 5 детьми имеют маленький вес в выборке, поэтому их можно не учитывать при анализе. Вывод: при наличии детей риск по невыплатам возрастает на 2%.

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [37]:
family_status_group =family_status_group.sort_values(by=['total'], ascending=False)
family_status_group

debt  total         % risk_group
family_status                                          
женат / замужем         931  12310  0.075630         no
гражданский брак        388   4154  0.093404        yes
Не женат / не замужем   274   2809  0.097544        yes
в разводе                85   1194  0.071189         no
вдовец / вдова           63    957  0.065831         no

**Вывод**

При рассмотрении категорий видно: те кто в браке или был в браке более ответсвенны по своим обязательствам: 6,5-7,5% невозврата, тогда как те кто не были в браке или живут в гражданском браке 9,3-9,7% не выполняют свои обязательства. Вывод: не бывавшие в браке клиенты на 2-3% чаще не испоолняют обязательства, риск выше.

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [38]:
income_type_group =income_type_group.sort_values(by=['total'], ascending=False)
income_type_group

debt  total         % risk_group
income_type                                    
низкий          479   5944  0.080585         no
ниже среднего   324   3870  0.083721        yes
средний         335   3870  0.086563        yes
выше среднего   329   3870  0.085013        yes
высокий         274   3870  0.070801         no

**Вывод**

Если рассматривать разбивку по доходу, то видно, что при высоких доходах % возврата выше, при маленьком и среднем доходе риск возрастает на 1,5% в среднем.

- Как разные цели кредита влияют на его возврат в срок?

In [39]:
category_purpose_group =category_purpose_group.sort_values(by=['total'], ascending=False)
category_purpose_group

debt  total         % risk_group
category_purpose                                  
недвижимость       964  12998  0.074165         no
образование        338   3559  0.094970        yes
автомобиль         317   3331  0.095167        yes
свадьба            122   1536  0.079427         no

**Вывод**

Если рассматривать с точки зрения цели кредита, то те клиенты, которые берут кредит на операции с недвижимостью и проведение свадьбы - более благонадежны, 7,4-7,9% соответственно, те же, кто берет кредит на образование или автомобиль чаще не исполняют свои обязательства на 2%.

## Шаг 4. Общий вывод

Для банка рекомендуется упростить порядок выдачи кредитов и сделать более лояльную систему для клиентов, кто уже в браке или бывал в браке, не имеет детей, с высоким доходом и берущим на операции с недвижимостью или на свадьбу. Риск по данным категориям группировки в среднем ниже на 2%.

Для клиентов, которые не были в браке, либо живут в гражданском браке, имеют детей, с низким или средним доходом, берущими кредит на цели по приобретению/ремонту автомобиля или образование рекомендуется разработать более тщательную систему кредитного скорринга для снижения рисков невозврата. В среднем данная группа чаще невыполняет свои обязательства 9,2-9,7%. 

<div class="alert alert-success">
<h2>Комментарий ревьюера</h2>

👍 Хорошая работа, можно дополнительно попробовать определить, является ли в целом группа клиентов группой риска, для этого надо сравнить процент должников в группе с общим процентом должников
</div>

<div class = "alert alert-info" >
<h2>Комментарий студента 1</h2>
Добавил по твоему совету столбик с определением группы риска.=)
</div>

<div class="alert alert-success">
<h2>Комментарий ревьюера V.2</h2>

👍 `sum / count` это `mean` (почти всегда)
</div>

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [X]  файл изучен;
- [X]  определены пропущенные значения;
- [X]  заполнены пропущенные значения;
- [X]  есть пояснение, какие пропущенные значения обнаружены;
- [X]  описаны возможные причины появления пропусков в данных;
- [X]  объяснено, по какому принципу заполнены пропуски;
- [X]  заменен вещественный тип данных на целочисленный;
- [X]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [X]  удалены дубликаты;
- [X]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [X]  описаны возможные причины появления дубликатов в данных;
- [X]  выделены леммы в значениях столбца с целями получения кредита;
- [X]  описан процесс лемматизации;
- [X]  данные категоризированы;
- [X]  есть объяснение принципа категоризации данных;
- [X]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [X]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [X]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [X]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [X]  в каждом этапе есть выводы;
- [X]  есть общий вывод.